### PR TITLE
REL-4170: GHOST Template

### PR DESCRIPTION
**Pull request by swalker2m**
Adds support for GHOST templates, according to the rules in `GHOST_BP.txt`.  The challenge with this template is that the resolution mode and target mode (taken together, the asterism type) is set at Phase 1 and we want to preserve this through instantiation.  Unfortunately this information is kept in the selected asterism in the target node, but there is no target node in the template group.  To compensate, a `preferredAsterismType` is added to the instrument component and overridden by GHOST in particular.  When there is no target node the instrument node's preferred asterism type is consulted instead.

### CODE REVIEW

File: bundle/edu.gemini.auxfile.workflow/src/main/java/edu/gemini/auxfile/copier/osgi/OsgiCopyConfig.java [hunk 1]
```diff
@@ -38,7 +41,9 @@ private OsgiCopyConfig(AuxFileType type, BundleContext ctx) {
 	    super(getProperty(type, ctx, HOST_KEY),
 	          getProperty(type, ctx, USER_KEY),
 	          getProperty(type, ctx, PASSWORD_KEY),
-              SshConfig$.MODULE$.DEFAULT_TIMEOUT());
+              SshConfig$.MODULE$.DEFAULT_TIMEOUT(),
+              ImOption.apply(ctx.getProperty(buildPropertyName(type, ENABLED_KEY))).forall("true"::equals)
```
Comment (swalker2m) [inline bundle/edu.gemini.auxfile.workflow/src/main/java/edu/gemini/auxfile/copier/osgi/OsgiCopyConfig.java:45]: I added an `enabled` bundle to property to make it possible to skip file sftps. This greatly simplifies testing. By default copying is enabled so it should not impact production.

File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala [hunk 2]
```diff
@@ -44,6 +44,6 @@ object Root {
 
   // REL-3769: GPI not available in 2020B.
   // GHOST not yet available
-  val choices: List[Instrument] = List(Alopeke, Flamingos2, /*Ghost,*/ GmosNorth, GmosSouth, Gnirs, Graces, Gsaoi, Igrins, MaroonX, Nifs, Niri, Visitor, Zorro)
+  val choices: List[Instrument] = List(Alopeke, Flamingos2, Ghost, GmosNorth, GmosSouth, Gnirs, Graces, Gsaoi, Igrins, MaroonX, Nifs, Niri, Visitor, Zorro)
```
Comment (swalker2m) [inline bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala:47]: Maybe we need to turn this back off for now?

File: bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/env/AsterismType.java [hunk 3]
```diff
@@ -91,24 +93,51 @@ public String displayValue() {
...
+
+    public static AsterismType forObservation(ISPObservation obs) {
+        return ImOption
+            .apply(SPTreeUtil.findTargetEnvNode(obs))
+            .map(n -> forTargetEnvironment(((TargetObsComp) n.getDataObject()).getTargetEnvironment()))
+            .orElse(ImOption
+                .apply(SPTreeUtil.findInstrument(obs))
+                .filter(n -> n.getDataObject() instanceof SPInstObsComp)
+                .map(n -> ((SPInstObsComp) n.getDataObject()).getPreferredAsterismType())
+            ).getOrElse(AsterismType.Single);
```
Comment (swalker2m) [inline bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/env/AsterismType.java:141]: This is where we figure out the asterism type, checking first the target environment and then falling back to the "preferred" type recorded in the instrument.

File: bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/template/TemplateGroup.java [hunk 4]
```diff
@@ -103,18 +63,6 @@ public void setGroupType(GroupType groupType) {
...
-    }
-
-    public void setStatus(Status status) {
-        if (this.status != status) {
-            final Status prev = this.status;
-            this.status = status;
-            firePropertyChange(PROP_STATUS, prev, status);
-        }
-    }
-
```
Comment (swalker2m) [inline bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/template/TemplateGroup.java:117]: `Status` was unused so I removed it.  In theory this should be fine.  😬

File: bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala [hunk 5]
```diff
@@ -190,4 +194,71 @@ object Asterism {
   def createSingleAsterism: Single = {
     Single(new SPTarget())
   }
+
+  def fromTypeAndTemplateTarget(
+    asterismType: AsterismType,
+    target:       SPTarget
+  ): Asterism = {
```
Comment (swalker2m) [inline bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala:201]: We have to create a proper GHOST asterism but have only one target from Phase 1.  This method compensates by adding and subtracting from the P1 target.  Presumably the PI will fill in the proper values during the phase 2 process after instantiation.

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/ghost/GhostEditor.scala [hunk 6]
```diff
@@ -628,6 +629,10 @@ final class GhostEditor extends ComponentEditor[ISPObsComponent, Ghost] {
       case SelectionChanged(`asterismComboBox`) => Swing.onEDT {
         val converter = asterismComboBox.selection.item.converter.asScalaOpt
         converter.foreach(convertAsterism)
+
+        val dataObj = getDataObject
+        dataObj.setPreferredAsterismType(asterismComboBox.selection.item)
+        getNode.setDataObject(dataObj)
```
Comment (swalker2m) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/ghost/GhostEditor.scala:635]: The editor now records the preferred asterism type as well.

File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala (hunk 2 above)
Comment (cquiroz) [reply bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala:47]: It was uncommented already for PIT 2023B

File: bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala (hunk 5 above)
Comment (cquiroz) [reply bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala:201]: Would it make sense to generate a second target for dual-target proposal to a given offset to the one in the proposal?

File: bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/obscomp/SPInstObsComp.java [hunk 7]
```diff
@@ -335,34 +336,6 @@ public void setParamSet(ParamSet paramSet) {
         }
     }
 
-    /**
-     * Are two SPInstObsComp objects are equal.
-     */
-    /*  equals / hashCode for mutable objects is not well defined and this
-        is problematic anyway because it is not defined in the subclasses
-    public boolean equals(Object obj) {
```
Comment (cquiroz) [inline bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/obscomp/SPInstObsComp.java:343]: Why was equals deleted?

File: bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala (hunk 5 above)
Comment (cquiroz) [reply bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala:201]: Oh, I see you're already doing that

File: bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala [hunk 8]
```diff
@@ -190,4 +194,71 @@ object Asterism {
...
+    asterismType match {
+      case AsterismType.Single                              =>
+        Single(target)
+
+      case AsterismType.GhostSingleTarget                   =>
+        GhostAsterism.SingleTarget(ghostTarget(target), None)
+
+      case AsterismType.GhostDualTarget                     =>
+        GhostAsterism.DualTarget(
+          ghostTarget(targetPlus(1)),
```
Comment (cquiroz) [inline bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/target/env/Asterism.scala:228]: I guess this means we generate two targets and the one from the proposal becomes the base at the center?

**Review by cquiroz — APPROVED** [review]
LGTM as far as the bits I can actually understand

File: bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala (hunk 2 above)
Comment (swalker2m) [reply bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala:47]: @cquiroz We now want to leave GHOST commented out correct?

Comment (cquiroz) [reply bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala:47]: Yes, it is commented out

Comment (bryanmiller) [reply bundle/edu.gemini.model.p1/src/main/scala/edu/gemini/model/p1/dtree/Root.scala]: We are going to want to release the GHOST P2 templates in the June release, supporting GHOST in the _2023A_ PIT. There will also probably be a new 2023B PIT release later in the year with GHOST available. I will be updating the GHOST template rules soon.

**Pull request merged**